File: Daily Resources/Day05/DSA - Prefix and Difference Arrays Solutions/forest_queries.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
    int n, q;
    cin >> n >> q;
    
    vector<string> forest(n);
    for (int i = 0; i < n; ++i) {
        cin >> forest[i];
    }

    vector<vector<int>> prefix(n + 1, vector<int>(n + 1, 0));

    // Build 2D prefix sum
    for (int y = 1; y <= n; ++y) {
        for (int x = 1; x <= n; ++x) {
            int is_tree = (forest[y - 1][x - 1] == '*') ? 1 : 0;
            prefix[y][x] = prefix[y - 1][x] + prefix[y][x - 1] - prefix[y - 1][x - 1]+ is_tree;
        }
    }

    // Process queries
    while (q--) {
        int y1, x1, y2, x2;
        cin >> y1 >> x1 >> y2 >> x2;
        int result = prefix[y2][x2]  - prefix[y1 - 1][x2] - prefix[y2][x1 - 1] + prefix[y1 - 1][x1 - 1];
        cout << result << '\n';
    }

    return 0;
}

File: Daily Resources/Day05/DSA - Prefix and Difference Arrays Solutions/forest_queries.py
n, q = map(int, input().split())

# Read forest grid
forest = [input().strip() for _ in range(n)]

# Build prefix sum grid
prefix = [[0] * (n + 1) for _ in range(n + 1)]

for y in range(1, n + 1):
    for x in range(1, n + 1):
        is_tree = 1 if forest[y - 1][x - 1] == '*' else 0
        prefix[y][x] = prefix[y - 1][x]+ prefix[y][x - 1]- prefix[y - 1][x - 1]+ is_tree
        

# Process queries
for _ in range(q):
    y1, x1, y2, x2 = map(int, input().split())
    result = prefix[y2][x2] - prefix[y1 - 1][x2] - prefix[y2][x1 - 1] + prefix[y1 - 1][x1 - 1]
    
    print(result)
